### ISSUE TITLE
Why is my model composed of GRU and ResNet training sooo much slower than their sum training times?

### ISSUE THREAD

**Issue by vdyagilev** [pytorch/pytorch#24822]
I have the following model, 
`class GRU_CNN(nn.Module):
    def __init__(self, num_features, batch_size, hidden_size):
        """Initialize the model by setting up the layers"""
        super(GRU_CNN, self).__init__()
        
        # initialize gru and cnn - the full models
        
        # gru model params
        self.num_features = num_features
        self.batch_size = batch_size
        self.hidden_size = hidden_size
        self.n_layers = 1
        
        # resnet model
        self.cnn = models.resnet34(pretrained=True, progress=False)
              
        # RNN-GRU model
        self.rnn = nn.GRU(input_size=self.num_features,
                          hidden_size=self.hidden_size)
        
        # init GRU hidden layer
        self.hidden = self.init_hidden(batch_size=self.batch_size, hidden_size=hidden_size)
        self.gru_output = nn.Linear(self.hidden_size, 1000)
        
        # final fully connected layers
        self.dense1 = nn.Linear(1000, 500)
        self.dense2 = nn.Linear(500, 100)
        self.dense3 = nn.Linear(100, 12)
        
        # output layer
        self.dense4 = nn.Linear(12, 1)
    
    def load_cnn_weights(self, cnn):
        cnn_params = cnn.named_parameters()
        gru_cnn_params = dict(self.cnn.named_parameters())
        
        for name, cnn_param in cnn_params:
            if name in gru_cnn_params:
                gru_cnn_params[name].data.copy_(cnn_param.data)
    
    def load_gru_weights(self, gru):
        gru_params = gru.named_parameters()
        gru_cnn_params = dict(self.rnn.named_parameters())
        
        for name, gru_param in gru_params:
            if name in gru_cnn_params:
                gru_cnn_params[name].data.copy_(gru_param.data)
    
    def forward(self, gru_input, cnn_input):
        """Perform a forward pass of our model on some input and hidden state"""
  
        # gru
        gru_out, self.hidden = self.rnn(gru_input, self.hidden)
        
        # detatch the hidden layer to prevent further backpropagating. i.e. fix the vanishing gradient problem
        self.hidden = self.hidden.detach().cuda()
        
        # pass through linear layer
        gru_out = torch.squeeze(self.gru_output(gru_out))
                
        # cnn
        cnn_out = self.cnn(cnn_input)
        
        # add the outputs of grunet and cnn
        x = gru_out.add(cnn_out)
        
        # feed through final layers

        # apply three fully-connected Linear layers with ReLU activation function
        x = self.dense1(x)
        x = relu(x)
        
        x = self.dense2(x)
        x = relu(x)
        
        x = self.dense3(x)
        x = relu(x)
        
        # output is a size 1 Tensor
        x = self.dense4(x)
        
        return x
    
    def init_hidden(self, batch_size, hidden_size):
        """Initializes hidden state"""
        
        # Creates initial hidden state for GRU of zeroes
        hidden = torch.ones(1, self.batch_size, hidden_size).cuda()
        return hidden`

which is a combination of a resnet34 and a GRU model. I sum their outputs and feed it through some more Linear layers to get my desired output. I also have two separate models containing just the resnet34 and GRU models with the 4 linear layers. When I train GRU and CNN my training speed is fairly short. However, when it took 5-10 seconds for a training epoch of batch size 64 for GRU and CNN, it takes 5 minutes for the same batch size for the combination model. I would like very much to hear if there are some feature optimizations for my code, or if backprop works slower on combination models. Any ideas are welcome. Thanks again.

Here is my training loop
`def train_dual(model, num_epochs, batch_size, train_gen1, train_gen2, valid_gen1, valid_gen2, test_gen, gru=False):
    """Standard training function used by all three models"""
    # For optimizing our model, we choose SGD 
    optimizer = optim.SGD(model.parameters(), lr=1e-3, momentum=0.9)
    
    # training loop
    
    # toop through the dataset num_epoch times
    for epoch in range(num_epochs):
        
        # train loop
        
        train_loss = []
        valid_loss = []
        
        # loop through each batch
        for i  in range(batch_size):
            gru_batch, gru_labels = next(iter(train_gen1))
            gru_batch, gru_labels = gru_batch.cuda(), gru_labels.cuda()
            gru_batch, gru_labels = gru_batch.float(), gru_labels.float()
            
            # add extra dimension to every vector in batch
            gru_batch.unsqueeze_(-1)
            gru_batch = gru_batch.expand(gru_batch.shape[0], gru_batch.shape[1], 1)

            # reformat dimensions
            gru_batch = gru_batch.transpose(2,0)
            gru_batch = gru_batch.transpose(1, 2)
            cnn_batch, cnn_labels = next(iter(train_gen2))
            cnn_batch, cnn_labels = cnn_batch.cuda(), cnn_labels.cuda()
            cnn_batch, cnn_labels = cnn_batch.float(), cnn_labels.float()
            
            # clear gradients
            model.zero_grad()
            output = model(gru_batch, cnn_batch)
            output = output[0]
            # declare the loss function and calculate output loss
            
            # we use the RMSE error function to train our model
            criterion = nn.MSELoss()
            
            loss = torch.sqrt(criterion(output, gru_labels))
            
            # backpropogate loss through model
            loss.backward()
            # perform model training based on propogated loss
            optimizer.step()
            
            train_loss.append(loss)
        
        
        # validation loop
        with torch.set_grad_enabled(False):
            for i in range(batch_size):
                gru_batch, gru_labels = next(iter(valid_gen1))
                gru_batch, gru_labels = gru_batch.cuda(), gru_labels.cuda()
                gru_batch, gru_labels = gru_batch.float(), gru_labels.float()
                
                # add extra dimension to every vector in batch
                gru_batch.unsqueeze_(-1)
                gru_batch = gru_batch.expand(gru_batch.shape[0], gru_batch.shape[1], 1)

                # reformat dimensions
                gru_batch = gru_batch.transpose(2,0)
                gru_batch = gru_batch.transpose(1, 2)

                cnn_batch, cnn_labels = next(iter(valid_gen2))
                cnn_batch, cnn_labels = cnn_batch.cuda(), cnn_labels.cuda()
                cnn_batch, cnn_labels = cnn_batch.float(), cnn_labels.float()
                
                # transform the model from training configuration to testing configuration. ex. dropout layers are removed
                model.eval()

                output = model(gru_batch, cnn_batch)
                output = output[0]
                
                val_loss = torch.sqrt(criterion(output, gru_labels))
                
                model.train()
                
                valid_loss.append(val_loss)
                
             # Profitability testing
            profit = 0.0
            
            for batch, labels in test_gen:
                gru_batch, gru_labels = next(iter(valid_gen1))
                gru_batch, gru_labels = gru_batch.cuda(), gru_labels.cuda()
                gru_batch, gru_labels = gru_batch.float(), gru_labels.float()
                
                # add extra dimension to every vector in batch
                gru_batch.unsqueeze_(-1)
                gru_batch = gru_batch.expand(gru_batch.shape[0], gru_batch.shape[1], 1)

                # reformat dimensions
                gru_batch = gru_batch.transpose(2,0)
                gru_batch = gru_batch.transpose(1, 2)

                cnn_batch, cnn_labels = next(iter(valid_gen2))
                cnn_batch, cnn_labels = cnn_batch.cuda(), cnn_labels.cuda()
                cnn_batch, cnn_labels = cnn_batch.float(), cnn_labels.float()
                
                # transform the model from training configuration to testing configuration. ex. dropout layers are removed
                model.eval()

                output = model(gru_batch, cnn_batch)
                output = output[0]
                
                # if output is > 0 ==> model predict positive growth for the next five cycles. Purchase now and sell in 5 periods.
                
                for i, pred in enumerate(output):
                    if pred > 1: # price will increase
                        profit += labels[i]
                
                
                model.train()
                
        print("Epoch: {}/{}...".format(epoch+1, num_epochs),
              "Training Loss: {}".format(round(float(sum(train_loss)/len(train_loss)), 4)),
              "Validation Loss: {}".format(round(float(sum(valid_loss)/len(valid_loss)), 4)),
              "Profitability: {}".format(round(float(profit), 3)))            `

**Comment by cpuhrsch** [comment]
Please use https://discuss.pytorch.org (https://discuss.pytorch.org/) for questions. Please kindly feel encouraged to reopen this issue, if this does not apply.
